pg-extend/src/pg_datum.rs

```diff
@@ -80,22 +81,41 @@ pub trait TryFromPgDatum<'s>: Sized {
         memory_context: &'mc PgAllocator,
         datum: PgDatum<'mc>,
     ) -> Result<Self, &'static str>
+    where
+        Self: 's,
+        'mc: 's,
+    {
+        if let Some(datum) = datum.0 {
+            unsafe {
+                Self::try_cast(memory_context, datum)
+            }
+        } else {
+            Err("datum was NULL")
+        }
+    }
+
+    /// Used to perform the actual casting. You should not use this directly, but
+    /// you should implement it. `datum` is expected to be non-null
+    #[doc(hidden)]
+    unsafe fn try_cast<'mc>(
```

Comment:
I feel like we could add a new trait, say `PgCastable` that could be implemented for all i32->i64 sized types. Then we could exploit that the Datum isn't an indirect pointer, and is actually carrying the value with generic functions.